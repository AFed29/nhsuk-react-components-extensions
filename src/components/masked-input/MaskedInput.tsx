import React, { HTMLProps, useState, useEffect, RefObject } from 'react';
import { useFormContext, Label, ErrorMessage, Hint } from 'nhsuk-react-components';
import InputMask, { ReactInputMask } from 'react-input-mask';
import classNames from 'classnames';

type MaskedInputProps = HTMLProps<HTMLInputElement> & {
  labelProps?: HTMLProps<HTMLLabelElement>;
  hint?: string;
  error?: boolean | string;
  width?: '2' | '3' | '4' | '5' | '10';
  mask: string;
  maskChar?: string;
  formatChars?: { [character: string]: string };
  alwaysShowMask?: boolean;
  inputRef?: (instance: HTMLInputElement | null) => any;
  ref?: string | ((instance: ReactInputMask | null) => void) | RefObject<ReactInputMask> | null;
};

const MaskedInput: React.FC<MaskedInputProps> = ({
  className,
  children,
  id,
  labelProps,
  hint,
  width,
  error,
  ...rest
}) => {
  const { isForm, setError } = useFormContext();
  const [name] = useState<string>(
    rest.name || `masked-input_${(Math.random() + 1).toString(36).substring(2, 7)}`,
  );

  useEffect(() => {
    if (isForm) {
      setError(name, Boolean(error));
    }
  }, [error, name, isForm]);

  return (
    <>
      {children ? (
        <Label
          className={labelProps ? labelProps.className : undefined}
          htmlFor={id}
          id={id ? `${id}-label` : undefined}
        >
          {children}
        </Label>
      ) : null}
      {hint ? <Hint id={id ? `${id}-hint` : undefined}>{hint}</Hint> : null}
      {error && typeof error === 'string' ? (
        <ErrorMessage id={id ? `${id}-error` : undefined}>{error}</ErrorMessage>
      ) : null}
      <InputMask
        className={classNames(
          'nhsuk-input',
          { [`nhsuk-input--width-${width}`]: width },
          { 'nhsuk-input--error': error },
          className,
        )}
        aria-describedby={children && id ? `${id}-label` : undefined}
        {...rest}
      />
    </>
  );
};

MaskedInput.defaultProps = {
  type: 'text',
};

export default MaskedInput;
