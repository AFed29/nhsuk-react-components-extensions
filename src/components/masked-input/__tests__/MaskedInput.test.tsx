import React from 'react';
import { mount } from 'enzyme';
import MaskedInput from '../MaskedInput';

describe('MaskedInput', () => {
  it('matches snapshot', () => {
    const component = mount(<MaskedInput mask="999 999 9999">NHS Number</MaskedInput>);
    expect(component).toMatchSnapshot();
    expect(component.text()).toBe('NHS Number');
    component.unmount();
  });
});
