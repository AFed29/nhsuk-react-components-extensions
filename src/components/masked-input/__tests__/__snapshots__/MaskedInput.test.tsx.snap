// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaskedInput matches snapshot 1`] = `
<MaskedInput
  mask="999 999 9999"
  type="text"
>
  <Label>
    <BaseLabel>
      <label
        className="nhsuk-label"
      >
        NHS Number
      </label>
    </BaseLabel>
  </Label>
  <InputElement
    className="nhsuk-input"
    mask="999 999 9999"
    type="text"
  >
    <input
      className="nhsuk-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onPaste={[Function]}
      type="text"
    />
  </InputElement>
</MaskedInput>
`;
